uvicorn/protocols/websockets/wsproto_impl.py

```diff
@@ -269,49 +273,94 @@ async def send(self, message: "ASGISendEvent") -> None:
+                    self.close_sent = True
+                    event = events.RejectConnection(status_code=403, headers=[])
+                    output = self.conn.send(event)
                     self.transport.write(output)
+                    self.transport.close()
 
-            elif message_type == "websocket.close":
-                self.queue.put_nowait({"type": "websocket.disconnect", "code": 1006})
-                self.logger.info(
-                    '%s - "WebSocket %s" 403',
-                    self.scope["client"],
-                    get_path_with_query_string(self.scope),
-                )
-                self.handshake_complete = True
-                self.close_sent = True
-                event = events.RejectConnection(status_code=403, headers=[])
-                output = self.conn.send(event)
-                self.transport.write(output)
-                self.transport.close()
+                elif message_type == "websocket.http.response.start":
+                    self.response_started = True
+                    message = typing.cast("WebSocketResponseStartEvent", message)
+                    self.logger.info(
+                        '%s - "WebSocket %s" %i',
+                        self.scope["client"],
+                        get_path_with_query_string(self.scope),
+                        message["status"],
+                    )
+                    event = events.RejectConnection(
+                        status_code=message["status"],
+                        headers=message["headers"],
+                        has_body=True,
+                    )
+                    output = self.conn.send(event)
+                    self.transport.write(output)
 
+                else:
+                    msg = (
+                        "Expected ASGI message 'websocket.accept', 'websocket.close' "
+                        "or 'websocket.http.response.start' "
+                        "but got '%s'."
+                    )
+                    raise RuntimeError(msg % message_type)
             else:
-                msg = (
-                    "Expected ASGI message 'websocket.accept' or 'websocket.close', "
-                    "but got '%s'."
-                )
-                raise RuntimeError(msg % message_type)
+                if message_type == "websocket.http.response.body":
+                    message = typing.cast("WebSocketResponseBodyEvent", message)
+                    body_finished = not message.get("more_body", False)
+                    event = events.RejectData(
+                        data=message["body"], body_finished=body_finished
+                    )
+                    output = self.conn.send(event)
+                    self.transport.write(output)
+                    if body_finished:
+                        self.queue.put_nowait(
+                            {"type": "websocket.disconnect", "code": 1006}
+                        )
```

Comment:
Why would you send this message?

Reply:
https://github.com/django/asgiref/issues/381

Reply:
What does `hypercorn` do? Let's follow their decision.

Reply:
Let's wait for Andrew's reply on the issue above, but to me, you shouldn't expect a message when sending a response. :thinking:

Reply:
Andrew replied. It's good to send the disconnect message. :+1: